worker/download.go

```diff
@@ -0,0 +1,1088 @@
+	resetTimer = func() {
+		timer.Stop()
+		select {
+		case <-timer.C:
+		default:
+		}
+		timer.Reset(s.mgr.overdriveTimeout)
+	}
+
+	// create a function to check whether overdrive is possible
+	canOverdrive := func() bool {
+		s.mu.Lock()
+		defer s.mu.Unlock()
+
+		// overdrive is not due yet
+		if time.Since(s.lastOverdrive) < s.mgr.overdriveTimeout {
+			return false
+		}
+
+		// overdrive is maxed out
+		remaining := s.minShards - s.numCompleted
+		if s.numInflight >= s.mgr.maxOverdrive+uint64(remaining) {
+			return false
+		}
+
+		s.lastOverdrive = time.Now()
+		return true
+	}
+
+	// try overdriving every time the timer fires
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-timer.C:
+				if canOverdrive() {
+					req := s.nextRequest(ctx, respChan, true)
+					if req == nil {
+						fmt.Printf("DEBUG PJ: %v | %d | no sector to overdrive\n", s.dID, s.index)
+					} else {
+						lErr := s.launch(req) // ignore error
+						fmt.Printf("DEBUG PJ: %v | %d | overdriving sector %d err %v\n", s.dID, s.index, req.sectorIndex, lErr)
+					}
+				}
+				resetTimer()
+			}
+		}
+	}()
+
+	return
+}
+
+func (s *slabDownload) nextRequest(ctx context.Context, responseChan chan sectorDownloadResp, overdrive bool) *sectorDownloadReq {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// prepare next sectors to download
+	if len(s.hostToSectors[s.curr]) == 0 {
+		// grab unused hosts
+		var hosts []types.PublicKey
+		for host := range s.hostToSectors {
+			if _, used := s.used[host]; !used {
+				hosts = append(hosts, host)
+			}
+		}
+
+		// sort them and assign the current host
+		s.mgr.sort(hosts)
+		for _, host := range hosts {
+			s.used[host] = struct{}{}
+			s.curr = host
+			break
+		}
+
+		// no more sectors to download
+		if len(s.hostToSectors[s.curr]) == 0 {
+			return nil
+		}
+	}
+
+	// pop the next sector
+	sector := s.hostToSectors[s.curr][0]
+	s.hostToSectors[s.curr] = s.hostToSectors[s.curr][1:]
+
+	// create the span
+	sCtx, span := tracing.Tracer.Start(ctx, "sectorDownloadReq")
+	span.SetAttributes(attribute.Stringer("hk", sector.Host))
+	span.SetAttributes(attribute.Bool("overdrive", overdrive))
+	span.SetAttributes(attribute.Int("sector", sector.index))
+
+	// build the request
+	return &sectorDownloadReq{
+		ctx: sCtx,
+
+		offset: s.offset,
+		length: s.length,
+		root:   sector.Root,
+		hk:     sector.Host,
+
+		overdrive:    overdrive,
+		sectorIndex:  sector.index,
+		responseChan: responseChan,
+	}
+}
+
+func (s *slabDownload) downloadShards(ctx context.Context, nextSlabTrigger chan struct{}) ([][]byte, error) {
+	// cancel any sector downloads once the download is done
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// add tracing
+	ctx, span := tracing.Tracer.Start(ctx, "downloadShards")
+	defer span.End()
+
+	// create the response channel
+	respChan := make(chan sectorDownloadResp)
+
+	// launch overdrive
+	resetOverdrive := s.overdrive(ctx, respChan)
```

Comment:
I understand that you want both upload and download code to be consistent which makes this a little odd. Because `overdrive` in the upload code actually overdrives and returns a request while this returns a trigger to reset the overdrive while having the same name.

Reply:
I originally wanted to extract the overdrive code from the upload code too, but didn't because I needed to reset it at one point in the main loop. I think the timer code and overdrive goroutine makes the main function overall less readable. So it's actually the other way around wrt consistency, it's not the download manager that's inconsistent, it's that I forgot to make the upload manager, which I now did in #433